pax_achievements.py

```diff
@@ -1,434 +1,258 @@
+    x = (
+        df.assign(year=df.date.dt.year)
+        .query("(q_flag == 1) and not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao_id": "count", "date": "max"})
+    )
+    return x[x.ao_id >= 20].reset_index().drop("ao_id", axis=1).rename({"date": "date_awarded"}, axis=1)
+
+
+def posts(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["year", "slack_user_id", "home_region"]
+    x = (
+        df.assign(year=df.date.dt.year)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao_id": "count", "date": "max"})
+    )
+    return x
+
+
+def six_pack(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["week", "slack_user_id", "home_region"]
+    x = (
+        df.assign(week=df.date.dt.isocalendar().week)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao_id": "count", "date": "max"})
+    )
+    return x[x.ao_id >= 6].reset_index().drop("ao_id", axis=1).rename({"date": "date_awarded"}, axis=1)
+
+
+def hdtf(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["year", "slack_user_id", "home_region", "ao_id"]
+    x = (
+        df.assign(year=df.date.dt.year)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao": "count", "date": "max"})
+    )
+    return x[x.ao >= 50].reset_index().drop("ao", axis=1).rename({"date": "date_awarded"}, axis=1)
+
+
+def load_to_database(row, engine: Engine, metadata: MetaData, data_to_load: pd.DataFrame) -> None:
+    """after successfully sending Slack notifications, push the data to the `achievements_awarded` table.
+    The data frame data_to_load has already been filtered to include only new achievements."""
+    try:
+        aa = metadata.tables[f"{row.paxminer_schema}.achievements_awarded"]
+    except KeyError:
+        aa = Table("achievements_awarded", metadata, autoload_with=engine, schema=row.paxminer_schema)
+
+    load_records = data_to_load.to_dict("records")
+    sql = insert(aa).values(load_records)
+    with engine.begin() as cnxn:
+        cnxn.execute(sql)
+
+
+def main():
+    year = date.today().year
+    engine = mysql_connection()
+    metadata = MetaData()
+    metadata.reflect(engine, schema="weaselbot")
+
+    df_regions = pd.read_sql(region_sql(metadata), engine)
+    nation_df = pd.read_sql(nation_sql(metadata, year), engine, parse_dates="date")
+
+    # for QSource, we want to capture only QSource
+    bb_filter = nation_df.backblast.str.lower().str[:100].str.contains(r"q.{0,1}source|q{0,1}[1-9]\.[0-9]\s")
+    ao_filter = nation_df.ao.str.contains(r"q.{0,1}source")
+
+    dfs = []
+    ############# Q Source ##############
+    dfs.append(the_priest(nation_df, bb_filter, ao_filter))
+    dfs.append(the_monk(nation_df, bb_filter, ao_filter))
```

Comment:
I know mine doesn't have this yet, but I want to start using the json field to capture event types for regions that are using custom fields. If enabled, the json field will contain a key for `"Event Type"` which could contain these options `["Bootcamp", "QSource", "Rucking", "2nd F"]` (which is the default options, though they can add or remove options). Logically, I would want this to overrule any of our other filter rules (so an event of "bootcamp" treats it as a standard beatdown, even if qsource or rucking is found in the title). Sorry if this makes this a lot more complicated!

Reply:
I wouldn't worry about this being complex. I just need to see it in action to make sure my understanding is accurate. I _think_ it's just a pivot from using the `backblast` field to the `json` field. Python will default to seeing that as a `dict` which is easy enough to manage. No need to muck around with `arrow` (yet?) until that time comes. Let me know your vision and when it's going to start to trickle down and I'll be glad to update the code to reflect using the `json` field.

Reply:
I think it makes sense at some point for us to make the event type field mandatory in Slackblast, then we don't have to mess around with this stuff; regions can be very clear with what they want to include vs. not. But again, until that time this is your call :)

BTW, here's what that `json` field will look like when populated: `{"Event Type": "Bootcamp"}`